### ISSUE TITLE
[2.8.8] Every services and hosts goes to Unknown status

### ISSUE THREAD

**Issue by Tomszy** [centreon/centreon#5609]
---------------------------------------------------
BUG REPORT INFORMATION
---------------------------------------------------

**Centreon Web version**: 2.8.8

**Centreon Engine version**: 1.7.2

**Centreon Broker version**: 3.0.8

**OS**: Oracle Linux 7.3

**Additional environment details (AWS, VirtualBox, physical, etc.):** Virtual

**Steps to reproduce the issue:**
1. Sometime every tests and hosts goes to Unknown status
2. No error in log files.
Fixing:
Need 1 or 2 cbd and centengine restart and the checks will come back

**Describe the results you received:**
![image](https://user-images.githubusercontent.com/25457811/29810962-20245f08-8ca2-11e7-95d4-9f822d6ee512.png)

**Describe the results you expected:**
It should be good if it will be never happen in the future.

**Comment by Sims24** [comment]
Hi, 

NTP and time sync are probably the root cause. Otherwise, check instance_timeout tag value in your central-broker.xml file.

**Comment by Tomszy** [comment]
Hi,

I don't find any instance_timeout tag in the xml file. Where can I find it in the interface to setup it?
How can i check if there is NTP and time sync problem?

**Comment by zskarman** [comment]
this should be similar issue.
[BUG] Poller/Hang or Stuck making all Host/Service Unknow after some time. #5407

**Comment by zskarman** [comment]
it may another similar issue..https://github.com/centreon/centreon-broker/issues/106

**Comment by lpinsivy** [comment]
@Tomszy you can check if ntp service is installed and in running state on all of your servers

https://www.cyberciti.biz/faq/howto-install-ntp-to-synchronize-server-clock/

**Comment by Tomszy** [comment]
Yes it is configured on every servers.

**Comment by Tomszy** [comment]
What should be a good value for this option? instance_timeout
We have nothing to setup for instance_timeout option.

**Comment by lpinsivy** [comment]
Did you update to Centreon Broker 3.0.8?

Problem persist?

**Comment by Sims24** [comment]
![image](https://user-images.githubusercontent.com/12429463/29866339-6b0a2940-8d78-11e7-92ec-65625f4282a1.png)

You find this option in your main broker. He is responsible of tagging outdated result as unknown/unreachable. 

Nevertheless, I think that you're problem comes from link loss from your poller broker to your central broker because of some data corruption in the BBDO protocol. YOu should consider upgrade central and pollers to centreon-broker 3.0.8 to fix this.

**Comment by Tomszy** [comment]
We have upgraded to 3.0.8 some minutes ago. Should we setup the instance_timeout option? And if yes then what should we setup? 120 sec?

**Comment by Sims24** [comment]
@Tomszy default is 300 sec and it's prelly good. Do not override it unless necessary.

**Comment by lpinsivy** [comment]
Ok I close this issue. Please reopen it if the problem appears again.

**Comment by Tomszy** [comment]
It was happened again 10 mins ago. Every services went to unknown status. We are using broker 3.0.8.
There is no error in log files.

**Comment by ganoze** [comment]
@Tomszy could you generate a debug file of your central-broker-master (with all message types and highly detailed verbosity) ?

**Comment by Tomszy** [comment]
Do you know the cli command to create this debug file?

**Comment by ganoze** [comment]
You can configure debug in Centreon, in the Configuration -> Pollers -> Broker configuration menu. There select central-broker-master and go to the Log tab. Enable all information types (debug, error, info and config) with a highly detailed verbosity. Save and restart the central poller, wait for the problem to reproduce (transition of all services to UNKNOWN state) and the log file should contain enough information.

**Comment by Tomszy** [comment]
Sorry. We cannot do this. We have 30 million SQL insert per day into the database and we cannot turn on debug logging, because this issue happen once or twice a week and If we turn on debug logging our disk on broker server will fill up quickly.
Anything else what we can do?

Or can you send us the sql query which count the Unknown number on the interface? And we need the php file as well where it is located. It should be a good starting to debug this issue.

**Comment by RaphNQ** [comment]
Hi !
I had the same problem (I have the same configuration) and I found a solution. The issue came from the notification. In my case, I've checked the debug log: Each time a notification was "sent" by the notification center, hosts changed to the following status: UNKNOWN. I have installed Postfix and changed the MTA by default (alternatives --config mta). If you want to check this solution, go to the users configuration and uncheck all users. Restart Centreon Engine and check.

**Comment by bouda1** [comment]
I have tried something following your idea, but to reproduce systematically the problem. I create a notification command that makes an infinite loop. And I put a service in hard critical state. Even a such command does not put all the services in unknown.

Have you seen the problem since your change ?

Do you still have your debug file ? Is it possible to have a look at it ?

Thanks.

**Comment by zskarman** [comment]
Hi,

The issue still exists.
What i found on the centreon_storage.hosts -> state is 2 similar for services as well.

Which process can do this like update centreon_storage.hosts SET state='2' ?
whithout any WHERE option?

**Comment by zskarman** [comment]
I found what code can do:
https://github.com/centreon/centreon-broker/tree/master/sql/sr/stream.cc
void stream::_update_hosts_and_services_of_instance(
               unsigned int id,
               bool responsive) {
  bool db_v2(_db.schema_version() == database::v2);
  std::ostringstream ss;
  if (responsive) {
    ss << "UPDATE " << (db_v2 ? "instances" : "rt_instances")
       << "  SET outdated=FALSE"
       << "  WHERE instance_id=" << id;
    database_query q(_db);
    q.run_query(ss.str(), "SQL: could not restore outdated instance");
    ss.str("");
    ss.clear();
    ss << "UPDATE " << (db_v2 ? "hosts" : "rt_hosts") << " AS h"
       << "  LEFT JOIN " << (db_v2 ? "services" : "rt_services")
       << "    AS s"
       << "    ON h.host_id=s.host_id"
       << "  SET h.state=h.real_state,"
       << "      s.state=s.real_state"
       << "  WHERE h.instance_id = " << id;
    q.run_query(ss.str(), "SQL: could not restore outdated instance");
  }
  else {
    ss << "UPDATE " << (db_v2 ? "instances" : "rt_instances")
       << "  SET outdated=TRUE"
       << "  WHERE instance_id=" << id;
    database_query q(_db);
    q.run_query(ss.str(), "SQL: could not outdate instance");
    ss.str("");
    ss.clear();
    ss << "UPDATE " << (db_v2 ? "hosts" : "rt_hosts") << " AS h"
       << "  LEFT JOIN " << (db_v2 ? "services" : "rt_services")
       << "    AS s"
       << "    ON h.host_id=s.host_id"
       << "  SET h.real_state=h.state,"
       << "      s.real_state=s.state,"
       << "      h.state=" << HOST_UNREACHABLE << ","
       << "      s.state=" << STATE_UNKNOWN
       << "  WHERE h.instance_id=" << id;
    q.run_query(ss.str(), "SQL: could not outdate instance");
  }
}

This can do 

UPDATE " << (db_v2 ? "hosts" : "rt_hosts") << " AS h"
       << "  LEFT JOIN " << (db_v2 ? "services" : "rt_services")
       << "    AS s"
       << "    ON h.host_id=s.host_id"
       << "  SET h.real_state=h.state,"
       << "      s.real_state=s.state,"
       << "      h.state=" << HOST_UNREACHABLE << ","
       << "      s.state=" << STATE_UNKNOWN

**Comment by bouda1** [comment]
I found something that can explain this problem.

If you create a very long notification command, engine tries to execute it during 30s. But if several users are notified with a such command, you cumulate 30s for each one, the delay can be sufficiently long so that broker believes there is "nothing in the pipe". Thus, hosts are marked as unreachable and services unknown.

So, a solution is to verify that notifications commands executed by centreon-engine do not finish in timeout. It is written in the centreon-engine.log file. Otherwise, attempt to optimize them or, worst solution, execute them in background.

**Comment by zskarman** [comment]
Hi,

So means, we have approx 40 pollers, and if any of the poller executed notif commands, which is goes to timeout, than all of the hosts are marked as unreachable and services unknown .

Rgs, Zsolt

**Comment by bouda1** [comment]
If notified users are sufficiently numerous and one notification goes to timeout for one poller, then all of the hosts  and services supervised by this poller are set to unreachable/unknown.

**Comment by lpinsivy** [comment]
@bouda1 thank you for your feedback

**Comment by lpinsivy** [comment]
Hi,

The problem of Unknown services comes from too long execution of notification command.
The execution of notification command is a blocking process to ensure that notification is sent.
This can delay the send of data to Centreon Broker.

Centreon Broker have a processus to set UNKNOWN services of a poller if data are not up to date from X minutes.

So that is why you can have UNKNOWN services in Centreon Web interface.

The best way is to optimize notification command and/or use "&" at the end of notification command.

Regards,

**Comment by zskarman** [comment]
Hi,

Could you please give me an example for: "&" at the end of notification command."

Regards, Zsolt

**Comment by zskarman** [comment]
Hi,

Now I got what is the & command.

And I have one question, if the issue is happen at one of the poller in my system, why is affect all of the services and hosts in the UI?

like all hosts  +services goes to unknown?

Also is that timeout is available on the poller? in the engine or broker log?

Regards, Zsolt

**Comment by zskarman** [comment]
Hi,,

For the testing I've created an Engine+DB and separated poller, and added 500 sec delay into the script what is handle the notification to simulate the timeout, But it looks nothing is happened.

But, if I killed the engine on the poller, than changed the centreon_storage.instantes last_alive value, changed to the past, then wait some minutes, then all of the host+checks goes to unknown for just on the poller's hosts.

So i can reproduce. But our problem is, at the same time all of the hosts + checks goes unknown.

My question is, how is it possible from broker perspective to get an insert to change all of the poller's last_alive value set to the past at the same time, than this should be cause our issue.

Regards, Zsolt
